File: src/libConfigurationManagement/src/PowerInfoManager.cpp
/*
 * Chris Bartley (bartley@cmu.edu)
 */

#include "PowerInfoManager.h"

const unsigned int PowerInfoManager::NUM_BREAKERS = 5;

const int PowerInfoManager::getMainBatteryVoltage()
   {
   return getHardwareProperty(QP_PROP_MAIN_BATT_VOLTAGE);
   }

const int PowerInfoManager::getBackupBatteryVoltage()
   {
   return getHardwareProperty(QP_PROP_BACKUP_BATT_VOLTAGE);
   }

const int PowerInfoManager::getIO5VVoltage()
   {
   return getHardwareProperty(QP_PROP_IO_5V_VOLTAGE);
   }

const int PowerInfoManager::getMain5VVoltage()
   {
   return getHardwareProperty(QP_PROP_MAIN_5V_VOLTAGE);
   }

const int PowerInfoManager::getMainBatteryState()
   {
   return getHardwareProperty(QP_PROP_MAIN_BATT_STATE);
   }

const int PowerInfoManager::getBackupBatteryState()
   {
   return getHardwareProperty(QP_PROP_BACKUP_BATT_STATE);
   }

const int PowerInfoManager::getIO5VState()
   {
   return getHardwareProperty(QP_PROP_IO_5V_STATE);
   }

const bool PowerInfoManager::getBreakerState(const unsigned int breakerId)
   {
   if (breakerId < NUM_BREAKERS)
      {
      const int breakersState = getHardwareProperty(QP_PROP_BREAKER_STATE);
      if (breakersState >= 0)
         {
         return (breakersState >> breakerId & 1) == 1;
         }
      }
   return false;
   }

Json::Value PowerInfoManager::getJSON()
   {
   Json::Value mainBatteryElement;
   mainBatteryElement["voltage"] = getMainBatteryVoltage();
   mainBatteryElement["state"] = getMainBatteryState();

   Json::Value backupBatteryElement;
   backupBatteryElement["voltage"] = getBackupBatteryVoltage();
   backupBatteryElement["state"] = getBackupBatteryState();

   Json::Value io5VElement;
   io5VElement["voltage"] = getIO5VVoltage();
   io5VElement["state"] = getIO5VState();

   Json::Value main5VElement;
   main5VElement["voltage"] = getMain5VVoltage();

   Json::Value breakersElement;
   for (unsigned int i=0; i<NUM_BREAKERS; i++)
      {
      breakersElement[i] = getBreakerState(i);
      }

   Json::Value powerInfoElement;
   powerInfoElement["main-battery"] = mainBatteryElement;
   powerInfoElement["backup-battery"] = backupBatteryElement;
   powerInfoElement["IO-5V"] = io5VElement;
   powerInfoElement["main-5V"] = main5VElement;
   powerInfoElement["breakers-state"] = breakersElement;

   Json::Value rootElement;
   rootElement["power-info"] = powerInfoElement;

   return rootElement;
   }

int PowerInfoManager::getHardwareProperty(const int property)
   {
   long val = -1;

   try
      {
      CQEPower &power = CQEPower::GetRef();
      if (power.GetProperty(property, &val) != PROP_OK)
         {
         // TODO: add some logging here
         }
      CQEPower::Release();
      }
   catch (...)
      {
      // TODO: add logging
      cerr << "PowerInfoManager::getHardwareProperty(): failed to get CQEPower reference required to get property." << endl;
      }

   return (int)val;
   }

File: src/www/javascript/edu/cmu/ri/terk/PowerInfoManager.js
//======================================================================================================================
// This file is part of Terk and TerkOS.
//
// All Terk and TerkOS source code is provided under the terms of the GNU General Public License v2
// (http://www.gnu.org/licenses/gpl-2.0.html). Those wishing to use Terk and TerkOS source code, software and/or
// technologies under different licensing terms should contact us at telepresencerobotkit@cs.cmu.edu. Such licensing
// terms are available for all portions of the Terk and TerkOS codebase presented here.
//======================================================================================================================
// Class for reading the hardware info.
//
// Dependencies:
// * jQuery (http://jquery.com/)
// * jQuery UI (http://jqueryui.com/)
//
// Author: Chris Bartley (bartley@cmu.edu)
//======================================================================================================================

//======================================================================================================================
// VERIFY NAMESPACE
//======================================================================================================================
// Create the global symbol "edu" if it doesn't exist.  Throw an error if it does exist but is not an object.
var edu;
if (!edu)
   {
   edu = {};
   }
else if (typeof edu != "object")
   {
   var eduExistsMessage = "Error: failed to create edu namespace: edu already exists and is not an object";
   alert(eduExistsMessage);
   throw new Error(eduExistsMessage);
   }

// Repeat the creation and type-checking for the next level
if (!edu.cmu)
   {
   edu.cmu = {};
   }
else if (typeof edu.cmu != "object")
   {
   var eduCmuExistsMessage = "Error: failed to create edu.cmu namespace: edu.cmu already exists and is not an object";
   alert(eduCmuExistsMessage);
   throw new Error(eduCmuExistsMessage);
   }

// Repeat the creation and type-checking for the next level
if (!edu.cmu.ri)
   {
   edu.cmu.ri = {};
   }
else if (typeof edu.cmu.ri != "object")
   {
   var eduCmuRiExistsMessage = "Error: failed to create edu.cmu.ri namespace: edu.cmu.ri already exists and is not an object";
   alert(eduCmuRiExistsMessage);
   throw new Error(eduCmuRiExistsMessage);
   }

// Repeat the creation and type-checking for the next level
if (!edu.cmu.ri.terk)
   {
   edu.cmu.ri.terk = {};
   }
else if (typeof edu.cmu.ri.terk != "object")
   {
   var eduCmuRiTerkExistsMessage = "Error: failed to create edu.cmu.ri.terk namespace: edu.cmu.ri.terk already exists and is not an object";
   alert(eduCmuRiTerkExistsMessage);
   throw new Error(eduCmuRiTerkExistsMessage);
   }
//======================================================================================================================

//======================================================================================================================
// DEPENDECIES
//======================================================================================================================
if (!window['$'])
   {
   var noJQuery = "The jQuery library is required by edu.cmu.ri.terk.PowerInfoManager.js";
   alert(noJQuery);
   throw new Error(noJQuery);
   }
//======================================================================================================================

//======================================================================================================================
// CODE
//======================================================================================================================
(function()
   {
   // ==================================================================================================================
   var jQuery = window['$'];
   jQuery.ajaxSetup({
      type: 'GET',
      dataType: 'jsonp',
      timeout: 10000,
      cache: false,
      global: false
   });

   edu.cmu.ri.terk.PowerInfoManager = function(host)
      {
      var eventListeners = new Array();
      var json = null;

      var notifyEventListeners = function(eventName)
         {
         jQuery.each(eventListeners, function(i, listener)
            {
            if (listener && listener[eventName])
               {
               listener[eventName]();
               }
            });
         };

      this.getPowerInfo = function()
         {
         notifyEventListeners('onBeforeLoad');

         // load the config
         jQuery.ajax(
         {
            url: host + '/cgi-bin/getPowerInfoAsJSON.pl',
            success: function(jsonResponse)
               {
               json = jsonResponse;

               if (json)
                  {
                  notifyEventListeners('onLoadSuccess');
                  }
               else
                  {
                  notifyEventListeners('onLoadFailure');
                  }
               },
            error: function()
               {
               notifyEventListeners('onLoadFailure');
               json = null;
               },
            complete: function()
               {
               notifyEventListeners('onLoadComplete');
               }
         });
         };

      /**
       * Registers the given event listener.  The event listener should implement some or all of the following methods:
       *
       *    onBeforeLoad()
       *    onLoadSuccess()
       *    onLoadFailure()
       *    onLoadComplete()
       */
      this.addEventListener = function(listener)
         {
         if (listener)
            {
            eventListeners[eventListeners.length] = listener;
            }
         };

      var getJSONProperty = function(jsonRoot, propertyPath)
         {
         if (jsonRoot)
            {
            if (propertyPath)
               {
               var propertyNames = propertyPath.split('.');
               for (var i = 0; i < propertyNames.length; i++)
                  {
                  var propertyName = propertyNames[i];
                  if (typeof jsonRoot[propertyName] == 'undefined')
                     {
                     return null;
                     }
                  else
                     {
                     jsonRoot = jsonRoot[propertyName];
                     }
                  }
               return jsonRoot;
               }

            return jsonRoot;
            }

         return null;
         };

      this.getMainBatteryVoltage = function()
         {
         return getJSONProperty(json, 'power-info.main-battery.voltage') / 1000;
         };

      this.getMainBatteryState = function()
         {
         return getJSONProperty(json, 'power-info.main-battery.state');
         };

      this.getBackupBatteryVoltage = function()
         {
         return getJSONProperty(json, 'power-info.backup-battery.voltage') / 1000;
         };

      this.getBackupBatteryState = function()
         {
         return getJSONProperty(json, 'power-info.backup-battery.state');
         };

      this.getIO5VVoltage = function()
         {
         return getJSONProperty(json, 'power-info.IO-5V.voltage') / 1000;
         };

      this.getIO5VState = function()
         {
         return getJSONProperty(json, 'power-info.IO-5V.state');
         };

      this.getMain5VVoltage = function()
         {
         return getJSONProperty(json, 'power-info.main-5V.voltage') / 1000;
         };

      this.getBreakersState = function()
         {
         return getJSONProperty(json, 'power-info.breakers-state');
         };
      // ---------------------------------------------------------------------------------------------------------------
      };

   // ==================================================================================================================
   })();
